src/main/scala/mesosphere/marathon/core/instance/Instance.scala

```diff
@@ -124,30 +135,100 @@ object Instance {
 
-      val task = Task(
-        taskId = Task.Id.forInstanceId(scheduledInstance.instanceId, None),
-        runSpecVersion = app.version,
-        status = Task.Status(
-          stagedAt = now,
-          condition = Condition.Created,
-          networkInfo = networkInfo
-        )
+      scheduledInstance.copy(
+        agentInfo = Some(agentInfo),
+        state = Instance.InstanceState(Condition.Provisioned, now, None, None),
+        tasksMap = Map(taskId -> Task(
+          taskId = taskId,
+          runSpecVersion = app.version,
+          status = Task.Status(
+            stagedAt = now,
+            condition = Condition.Created,
+            networkInfo = networkInfo
+          ))),
+        runSpecVersion = app.version
       )
+    }
+
+    /**
+      * Factory method for creating provisioned instance from Scheduled instance for pods
+      * @param scheduledInstance instance in a Scheduled state
+      * @return new instance in a provisioned state
+      */
+    def apply(
+      scheduledInstance: Instance,
+      agentInfo: Instance.AgentInfo,
+      hostPorts: Seq[Option[Int]],
+      pod: PodDefinition,
+      taskIds: Seq[Task.Id],
+      now: Timestamp): Instance = {
+      require(scheduledInstance.isScheduled, s"Instance '${scheduledInstance.instanceId}' must not be in state '${scheduledInstance.state.condition}'. Scheduled instance is required to create provisioned instance.")
 
-      val tasksMap = Map(task.taskId -> task)
+      val taskNetworkInfos = podTaskNetworkInfos(pod, agentInfo, taskIds, hostPorts)
 
       scheduledInstance.copy(
         agentInfo = Some(agentInfo),
         state = Instance.InstanceState(Condition.Provisioned, now, None, None),
-        tasksMap = tasksMap,
-        runSpecVersion = app.version
+        tasksMap = taskIds.map { taskId =>
+          // the task level host ports are needed for fine-grained status/reporting later on
+          val networkInfo = taskNetworkInfos.getOrElse(
+            taskId,
+            throw new IllegalStateException("failed to retrieve a task network info"))
+          val task = Task(
+            taskId = taskId,
+            runSpecVersion = pod.version,
+            status = Task.Status(stagedAt = now, condition = Condition.Created, networkInfo = networkInfo)
+          )
+          task.taskId -> task
+        }(collection.breakOut),
+        runSpecVersion = pod.version
       )
     }
   }
 
+  private def podTaskNetworkInfos(
+    pod: PodDefinition,
+    agentInfo: Instance.AgentInfo,
+    taskIDs: Seq[Task.Id],
+    hostPorts: Seq[Option[Int]]
+  ): Map[Task.Id, NetworkInfo] = {
+
+    val reqPortsByCTName: Seq[(String, Option[Int])] = pod.containers.flatMap { ct =>
+      ct.endpoints.map { ep =>
+        ct.name -> ep.hostPort
+      }
+    }
+
+    val totalRequestedPorts = reqPortsByCTName.size
+    assume(totalRequestedPorts == hostPorts.size, s"expected that number of allocated ports ${hostPorts.size}" +
```

Comment:
what do you think about using `require` here? otherwise `assume` can me disabled in runtime with `-Xdisable-assertions`